File: DFS/040.Combination-Sum-II/040.Combination-Sum-II.ts
function combinationSum2(candidates: number[], target: number): number[][] {
  candidates.sort((a, b) => a - b);
  const ret: number[][] = [];
  const current: number[] = [];
  dfs(0);
  return ret;

  function dfs(index: number, sum = 0) {
    if (sum > target) {
      return;
    }

    if (sum === target) {
      ret.push([...current]);
    }

    for (let i = index; i < candidates.length; i++) {
      if (i === index || candidates[i] !== candidates[i - 1]) {
        current.push(candidates[i]);
        dfs(i + 1, sum + candidates[i]);
        current.pop();
      }
    }
  }
}

File: DFS/040.Combination-Sum-II/040.Combination-Sum-II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        current = []
        
        def dfs(index, sum):
            if sum > target:
                return
            
            if sum == target:
                result.append(current[:])
                return

            for i in range(index, len(candidates)):
                if i == index or candidates[i] != candidates[i - 1]:
                    current.append(candidates[i])
                    dfs(i + 1, sum + candidates[i])
                    current.pop()
        dfs(0, 0)
        return result
